executor/runtime/docker/docker.go

```diff
@@ -2677,6 +2682,88 @@ stopped:
 	return err
 }
 
+// runAllPreStopHooks emulates k8s lifecycle preStop hook sequence
+// https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks
+// Hooks are blocking, and pods should not assume any particular order
+func (r *DockerRuntime) runAllPreStopHooks(ctx context.Context) *multierror.Error {
+	var errs *multierror.Error
+	for _, c := range r.c.ExtraPlatformContainers() {
+		err := r.runPreStopHookIfDefined(ctx, &c.V1Container, c.Status.ContainerID)
+		errs = multierror.Append(errs, err)
+	}
+	for _, c := range r.c.ExtraUserContainers() {
+		err := r.runPreStopHookIfDefined(ctx, &c.V1Container, c.Status.ContainerID)
+		errs = multierror.Append(errs, err)
+
+	}
+	err := r.runPreStopHookIfDefined(ctx, &r.c.Pod().Spec.Containers[0], r.c.ID())
+	errs = multierror.Append(errs, err)
+	return errs
+}
+
+func (r *DockerRuntime) runPreStopHookIfDefined(ctx context.Context, c *v1.Container, cid string) error {
+	if c.Lifecycle == nil {
+		return nil
+	}
+	if c.Lifecycle.PreStop == nil {
+		return nil
+	}
+	if cid == "" {
+		return fmt.Errorf("Unable to run any preStop hooks for %s container, cid is empty?", c.Name)
+	}
+	if c.Lifecycle.PreStop.HTTPGet != nil {
+		return fmt.Errorf("HTTP prestop hook on %s container not supported, ignoring", c.Name)
+	}
+	if c.Lifecycle.PreStop.TCPSocket != nil {
+		return fmt.Errorf("TCPSocket prestop hook on %s container not supported, ignoring", c.Name)
+	}
+	if c.Lifecycle.PreStop.Exec == nil {
+		return nil
+	}
+	if c.Lifecycle.PreStop.Exec.Command == nil {
+		return nil
+	}
+	return r.runBlockingPreStopExec(ctx, c.Lifecycle.PreStop.Exec.Command, cid, c.Name)
+}
+
+// runBlockingPreStopExec runs 'docker exec' in a container, for the purposes of running a preStop hook for a container
+// This is meant as a best effort sort of thing, where errors are reported, but callers of this function should just report
+// on them. Tini is also involved, because it difficult to debug these preStop hooks, so having the output automatically
+// redirected to a file in /logs allows users to see them after the fact.
+func (r *DockerRuntime) runBlockingPreStopExec(ctx context.Context, command []string, cid string, cName string) error {
```

Comment:
I'm not too familiar with that class of vulnerability.
The only thing a titus user has under their control is the command itself. Do those vulnerabilities involve crafting a command or something else?

Also keep in mind that this hook configuration is set by platform sidecar authors, not end users.

No, I don't think it should be in tini. We would have to have a socket, and it would be a sort of root exec as a service over a socket? Or I guess it could be an env var. But we can't 100% guarantee that tini will always be running. In main containers it is handed off to systemd, etc.

Reply:
I agree that this vuln is bad and we would need to upgrade titus if we saw this.
This PR is not the only case of 'exec' in action (titus system services also do this, and can load other libraries, run hooks, etc), so I don't think I'm adding too much security surface here.